### ISSUE TITLE
Making nuget work

### ISSUE THREAD

**Issue by zivkan** [Microsoft/VSProjectSystem#244]
## Repro Steps

1. Clone [my fork](https://github.com/zivkan/VSProjectSystem)
2. Open `samples/WindowsScript/WindowsScript.sln` and build the solution
3. Start the Visual Studio experimental instance (Run without debugging)
4. In the experimental instance, open `samples/project/sample/sample.sln`
5. Manage nuget packages for the `WindowsScript` project, and add any package. (the use-case is to add a reference to a nuget package with the msbuild targets and props, but that's not important to repo this issue. If you wish to try with a nuget that has build tools, add `Adfc.Msbuild`, or Service Fabric's msbuild package`)

Alternatively, you can clone this repo, and edit the sample by adding the `PackageReferences`, `AssemblyReferences`, `DeclaredSourceItems`, and `UserSourceItems` project capabilities. The sample `.wsproj` file was also edited to have a line `<Import Project="$(MSBuildToolsPath)\Microsoft.Common.targets" />`.

## Expected Behaviour

1. Visual Studio should add a `<PackageReference` item in the `.wsproj` file
2. Visual Studio should do a nuget restore, creating the appropriate files in the obj/ directory
3. ~~If the nuget package contains build targets and props that include file globs, I expect Visual Studio to show the files that match the globs in Solution Explorer. The `Adfc.Msbuild` package has a glob for `**\*.json`, so I would expect the `file.json` file in the WindowsScript directory to appear without selecting `Show All Files`~~

## Actual Behaviour

1. After adding the nuget package, a `packages.config` file is added, and the `.wsproj` file gets a project item to this file instead of a package reference.
2. Even if I manually edit the `.wsproj` file to add the PackageReference item, Visual Studio doesn't do a nuget restore, in order to create the props and targets files in the obj/ directory
3. ~~Even if I manually run `msbuild /t:restore` from the command prompt after adding the `Adfc.Msbuild` package, `file.json` is not shown in Solution Explorer.~~ I just needed the right xaml rules files.

**Comment by zivkan** [comment]
I forked this repo and modified it to create a minimally reproducible sample, if that helps. I also edited my first post to make it much clearer.

**Comment by jviau** [comment]
Hi @zivkan,

For the issue with showing `<Json Include="**/*.json>` you will need to first _tell_ CPS about that item type. This is done via the xaml rule system. Take a look in the sample [BuildSystem/Rules](https://github.com/Microsoft/VSProjectSystem/tree/master/samples/WindowsScript/WindowsScript/WindowsScript.ProjectType/BuildSystem/Rules) to see how this is done.

You will need a few main things:
* `ContentType`, `ItemType`, and `FileExtension`, like in [ProjectItemsSchema.xaml](https://github.com/Microsoft/VSProjectSystem/blob/master/samples/WindowsScript/WindowsScript/WindowsScript.ProjectType/BuildSystem/Rules/ProjectItemsSchema.xaml)
* `Rule`, such as [none.xaml](https://github.com/Microsoft/VSProjectSystem/blob/master/samples/WindowsScript/WindowsScript/WindowsScript.ProjectType/BuildSystem/Rules/none.xaml).
But also include this property in it, which our example is missing. It lets the user switch the item type via a drop down in VS:
``` xml
<DynamicEnumProperty Name="{}{ItemType}" DisplayName="Build Action"  Category="Advanced"
                   Description="How the file relates to the build and deployment processes."
                   EnumProvider="ItemTypes" />
```

* Finally, include those 2 .xaml files as `<PropertyPageSchema Include="..." Context="File;BrowseObject" />` in your .targets or .props. Such as in [WindowsScripts.targets](https://github.com/Microsoft/VSProjectSystem/blob/master/samples/WindowsScript/WindowsScript/WindowsScript.ProjectType/BuildSystem/DeployedBuildSystem/WindowsScript.targets)

**Comment by jviau** [comment]
Also, it is recommended to put globs in `.props`. This way CPS can make use of the msbuild `Update` element to selectively edit a globbed items metadata in the middle of the project file, and not have to add items below the .targets imports.

**Comment by zivkan** [comment]
@jviau thank you very much! I really appreciate it. The JSON files are appearing in Solution Explorer now. I remember when looking at the xaml files a few weeks ago that I thought I needed to create something for my json item type, but I got distracted with other issues and forgot about it. Without your help it would have taken me weeks of comparing dotnet/project-system's files to the sample here to find the DynamicEnumProperty, so thanks for telling me about it.

And thanks for the advice on putting my item globs in the props file. Before starting this project I had no experience with creating msbuild files or Visual Studio extensions, so it's been an enormous learning curve for me. I'll create a task to move the globs to the props file, so I don't forget the next time I publish a new nuget package.

Nearly ready for my first release! Just need PackageReference and nuget restore to work in Visual Studio (and figure out how to create an installer)

**Comment by jviau** [comment]
VSIX is typically sufficient for an installer now. The main issue CPS projects couldn't use VSIX was they always installed to a hash-d subfolder, so you couldn't reference your deployed targets. But in VS2017 support has been added to deploy some files to the msbuild subdirectory via a vsix, so you can place your targets in there and have no problem referencing them via MSBuild well known properties.

As for adding PackageReference and nuget restore, that is all owned by [.netcore project system](https://github.com/dotnet/project-system). I am not sure if it will work outside of their projects right now, but you can check their github for that.

**Comment by zivkan** [comment]
thanks again for your reply. You're right, automatic nuget restore of PackageReference is currently tied specifically to C#, VB and F# projects. The initial development appears to have intentionally excluded everyone else from being able to use it, as I discovered in issue 5868 in the nuget home repo, linked above. But now they have an issue to make it available for other project types. Sorry I forgot to update this issue.

However, the issue remains that despite having the PackageReferences capability defined in my project's target file, when I use Manage Nuget Packages to install a package, it creates a packages.config file, instead of a new itemgroup and item in my project file. Is this a CPS issue, or a nuget issue?  Perhaps the project capabilies doco that says this capability exists should be updated to tell people like me that it doesn't work for custom project types yet. I've wasted a lot of time trying to get it to work.

And despite creating a packages.config file, and the nuget package containing the expected props and targets file, the project file does not get updated to import those two files from my nuget. But installing the same nuget package in a csproj file (with packages.config) does, so I'm confident the package was authored correctly. However, this is clearly a nuget issue that I guess I should report there, not here.

My current workaround is to remove the capabilies that makes Manage Nuget Packages work on my project, and force users to manually edit the proj file as xml to change the package version (I hope to find a way to programatically change it in the future, and create my own UI, since my project only needs 1 nuget reference, never more). And I use IVsSolutionEvent5's OnBeforeOpenProject to detect when my project type's guid is being loaded, then programatically call msbuild to run the restore target. It means users have to close the solution and open it again to start using the new build tools, but I'm otherwise blocked.

Finally, it's fantastic news that vsix can install to msbuild. I think I see how to do it, I just need to test it now. Thank you very much for telling me about it. I wouldn't have discovered this functionality myself.
